### PR TITLE
first draft using utopia-php database

### PR DESCRIPTION
**Pull request by lohanidamodar**

### CODE REVIEW

File: src/Audit/Audit.php [hunk 1]
```diff
@@ -2,19 +2,47 @@
...
 
     /**
-     * @param Adapter $adapter
+     * @param Database $adapter
      */
-    public function __construct(Adapter $adapter)
+    public function __construct(Database $db)
+    {
+        $this->db = $db;
+        $this->init();
```
Comment (eldadfux) [inline src/Audit/Audit.php:24]: We don't have to trigger init on the constructor.

File: src/Audit/Audit.php [hunk 2]
```diff
@@ -2,19 +2,47 @@
...
+    }
+
+    private function init(): void
```
Comment (eldadfux) [inline src/Audit/Audit.php:27]: Double thinking about it, I think `install` or `setup` might better describe this method.

File: src/Audit/Audit.php [hunk 3]
```diff
@@ -2,19 +2,47 @@
...
     {
-        $this->adapter = $adapter;
+        if (!$this->db->exists()) {
+            $this->db->create();
+            $this->db->createCollection(Audit::COLLECTION);
+            $this->db->createAttribute(Audit::COLLECTION, 'userId', Database::VAR_STRING, 45, true);
```
Comment (eldadfux) [inline src/Audit/Audit.php:32]:
```suggestion
            $this->db->createAttribute(Audit::COLLECTION, 'userId', Database::VAR_STRING, Database::LENGTH_KEY, true);
```

File: src/Audit/Audit.php [hunk 4]
```diff
@@ -2,19 +2,47 @@
...
+            $this->db->createAttribute(Audit::COLLECTION, 'event', Database::VAR_STRING, 45, true);
+            $this->db->createAttribute(Audit::COLLECTION, 'resource', Database::VAR_STRING, 45, false);
+            $this->db->createAttribute(Audit::COLLECTION, 'userAgent', Database::VAR_STRING, 65534, true);
+            $this->db->createAttribute(Audit::COLLECTION, 'ip', Database::VAR_STRING, 45, true);
+            $this->db->createAttribute(Audit::COLLECTION, 'location', Database::VAR_STRING, 45, false);
+            $this->db->createAttribute(Audit::COLLECTION, 'time', Database::VAR_INTEGER, 0, true, false);
+            $this->db->createAttribute(Audit::COLLECTION, 'data', Database::VAR_STRING, 16777216, false,true,false,['json']);
```
Comment (eldadfux) [inline src/Audit/Audit.php:39]:
```suggestion
            $this->db->createAttribute(Audit::COLLECTION, 'event', Database::VAR_STRING, 255, true);
            $this->db->createAttribute(Audit::COLLECTION, 'resource', Database::VAR_STRING, 255, false);
            $this->db->createAttribute(Audit::COLLECTION, 'userAgent', Database::VAR_STRING, 65534, true);
            $this->db->createAttribute(Audit::COLLECTION, 'ip', Database::VAR_STRING, 45, true);
            $this->db->createAttribute(Audit::COLLECTION, 'location', Database::VAR_STRING, 45, false);
            $this->db->createAttribute(Audit::COLLECTION, 'time', Database::VAR_INTEGER, 0, true, false);
            $this->db->createAttribute(Audit::COLLECTION, 'data', Database::VAR_STRING, 16777216, false,true,false,['json']);
```

File: src/Audit/Audit.php [hunk 5]
```diff
@@ -34,7 +62,20 @@ public function __construct(Adapter $adapter)
...
+            'userId' => $userId,
+            'event' => $event,
+            'resource' => $resource,
+            'userAgent' => $userAgent,
+            'ip' => $ip,
+            'location' => $location,
+            'data' => $data,
+            'time' => \time()
+        ]));
+        return true;
```
Comment (eldadfux) [inline src/Audit/Audit.php:78]: Reset authorization back, leaving it disable could lead to a security bridge.

File: src/Audit/Audit.php [hunk 6]
```diff
@@ -2,19 +2,47 @@
...
 namespace Utopia\Audit;
 
+use Utopia\Database\Database;
+use Utopia\Database\Document;
+use Utopia\Database\Query;
+use Utopia\Database\Validator\Authorization;
+
 class Audit
 {
+    const COLLECTION = "abuse";
```
Comment (eldadfux) [inline src/Audit/Audit.php:12]:
```suggestion
    const COLLECTION = "audit";
```

File: tests/Audit/AuditTest.php [hunk 7]
```diff
@@ -36,18 +40,18 @@ public function setUp(): void
...
         $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);        // Handle all errors with exceptions
-
-        $adapter = new MySQL($pdo);
-
-        $adapter
-            ->setNamespace('namespace') // DB table namespace
-        ;
-
-        $this->audit = new Audit($adapter);
+
```
Comment (eldadfux) [inline tests/Audit/AuditTest.php:47]: Extra space

File: src/Audit/Audit.php [hunk 8]
```diff
@@ -2,19 +2,47 @@
...
+            $this->db->createAttribute(Audit::COLLECTION, 'userAgent', Database::VAR_STRING, 65534, true);
+            $this->db->createAttribute(Audit::COLLECTION, 'ip', Database::VAR_STRING, 45, true);
+            $this->db->createAttribute(Audit::COLLECTION, 'location', Database::VAR_STRING, 45, false);
+            $this->db->createAttribute(Audit::COLLECTION, 'time', Database::VAR_INTEGER, 0, true, false);
+            $this->db->createAttribute(Audit::COLLECTION, 'data', Database::VAR_STRING, 16777216, false,true,false,['json']);
+
+            $this->db->createIndex(Audit::COLLECTION, 'index_1', Database::INDEX_KEY, ['userId']);
+            $this->db->createIndex(Audit::COLLECTION, 'index_2', Database::INDEX_KEY, ['event']);
+            $this->db->createIndex(Audit::COLLECTION, 'index_3', Database::INDEX_KEY, ['resource']);
+
```
Comment (eldadfux) [inline src/Audit/Audit.php:44]: We can also add a shared index for userId and event

File: src/Audit/Audit.php [hunk 9]
```diff
@@ -73,18 +124,33 @@ public function getLogsByResource(string $resource): array
      */
     public function getLogsByUserAndActions(string $userId, array $actions): array
```
Comment (eldadfux) [inline src/Audit/Audit.php:125]: Let's fix the naming conventions. Action\s == Event\s

**Review by eldadfux — CHANGES_REQUESTED** [review]

File: src/Audit/Audit.php [hunk 10]
```diff
@@ -2,19 +2,47 @@
...
 
+use Utopia\Database\Database;
+use Utopia\Database\Document;
+use Utopia\Database\Query;
+use Utopia\Database\Validator\Authorization;
+
 class Audit
 {
+    const COLLECTION = "audit";
     /**
```
Comment (eldadfux) [inline src/Audit/Audit.php:13]:
```suggestion

    /**
```

**Comment by eldadfux** [comment]
We should also update the readme examples with the new changes

File: README.md [hunk 11]
```diff
@@ -22,31 +22,36 @@ Init the audit object:
 
 require_once __DIR__ . '/../../vendor/autoload.php';
 
+use PDO;
 use PDO;
 use Utopia\Audit\Audit;
-use Utopia\Audit\Adapters\MySQL;
+use PHPUnit\Framework\TestCase;
```
Comment (eldadfux) [inline README.md:28]:
```suggestion
```

File: README.md [hunk 12]
```diff
@@ -22,31 +22,36 @@ Init the audit object:
...
 
-$adapter
-    ->setNamespace('namespace') // DB table namespace
-;
+$database = new Database(new MySQL($pdo),$cache);
+$database->setNamespace('namespace');
 
-$audit = new Audit($adapter);
+$this->audit = new Audit($database);
+$this->audit->setup();
```
Comment (eldadfux) [inline README.md:54]:
```suggestion
$audit = new Audit($database);
$audit->setup();
```

**Review by eldadfux — CHANGES_REQUESTED** [review]

**Review by kodumbeats — CHANGES_REQUESTED** [review]
the `getLogs` methods state in their description to "get all logs", but without passing a limit, only the default limit of 25 will be returned.

**Comment by lohanidamodar** [comment]
> the `getLogs` methods state in their description to "get all logs", but without passing a limit, only the default limit of 25 will be returned.

It should be fine. I think we are using that only for tests.

**Review by eldadfux — APPROVED** [review]

File: tests/Audit/AuditTest.php [hunk 13]
```diff
@@ -15,39 +15,45 @@
...
+use Utopia\Database\Adapter\MySQL;
+use Utopia\Database\Database;
 
 class AuditTest extends TestCase
 {
     /**
      * @var Audit
      */
     protected $audit = null;
+    protected $initialized = false;
```
Comment (kodumbeats) [inline tests/Audit/AuditTest.php:30]: Is this variable used anywhere?

**Review by kodumbeats — CHANGES_REQUESTED** [review]
Minor question

**Comment by eldadfux** [comment]
Seems like we need to apply a default order for the logs in all `find()` methods using ['_id'] as the attribute and `DESC` as the order type. For some reason the `time` attribute also seems wrong.

image

**Comment by eldadfux** [comment]
I've found and fixed the date format problem on appwrite/appwrite, we just need to fix the logs ordering.

**Comment by eldadfux** [comment]
@lohanidamodar one last thing, I think we should change `getLogsByUserAndActions` to `getLogsByUserAndEvents`

**Comment by lohanidamodar** [comment]
> @lohanidamodar one last thing, I think we should change `getLogsByUserAndActions` to `getLogsByUserAndEvents`

Done!

**Pull request merged**